crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -678,12 +694,16 @@ impl Room {
     ///
     /// Returns the display names as a list of strings.
     async fn extract_heroes(&self, heroes: &[RoomHero]) -> StoreResult<Vec<String>> {
-        let own_user_id = self.own_user_id().as_str();
-
         let mut names = Vec::with_capacity(heroes.len());
-        let heroes = heroes.iter().filter(|hero| hero.user_id != own_user_id);
+        let own_user_id = self.own_user_id();
+        let member_hints = self.get_member_hints().await?;
+
+        // Construct a filter that is specific to this own user id, set of member hints,
+        // and accepts a `RoomHero` type.
+        let heroes_filter =
+            |hero: &&RoomHero| heroes_filter(own_user_id, &member_hints)(&hero.user_id);
```

Comment:
This may or may not create a the inner closure multiple times (once for every call to the outer `heroes_filter`). We can construct it once explicitly instead, with:

```rust
let parent_filter = heroes_filter(own_user_id, &member_hints);
let heroes_filter = |hero: &&RoomHero| parent_filter(&hero.user_id);
```

Reply:
a8827386ea6bc9462f2f8faffad49c7740096ac8